README.md
# Kerbal Space Program Automation

[![Build Status](https://travis-ci.org/prikhi/ksp-automation.svg?branch=master)](https://travis-ci.org/prikhi/ksp-automation)

A CLI KSP Automation program & library that utilizes the [kRPC][krpc] mod.

Currently in early alpha - the only current supported program is achieving
Low-Kerbin Orbit from the Kerbal Space Center Launchpad.

![The User Interface of the KSP Automation Program, Showing a Mission List, Parameters, Logs, and Options](http://bugs.sleepanarchy.com/projects/ksp-automation/repository/revisions/master/entry/screenshot.png "KSP Automation Screenshot")


## Goals

The future will include launch/abort buttons, more missions, tweakable
parameters for missions, global automation(like automatically staging whenever
fuel in stage is depleted).

Eventually this could be expanded into a "Mission Builder" where you can select
a vessel & tweak/assemble a list of tasks into larger missions. Something that
would let you create KSC -> LKO -> Dock at Kerbin Station -> Dump Fuel ->
Kerbin Landing. Some sort of management system for these custom missions.

A special interface for GHCi would be cool. I'd imagine something that hides
the connection & log arguments like:

    initializeConnection 50000 50001
    lowKerbinOrbit 80000
    setTarget "KerbStation"
    rendezvous
    activateStage
    reEnterKerbin


## Build / Run

There are currently no downloadable releases, you will need to build the
application yourself. The build process requires [Haskell stack][stack]:

    sudo pacman -S stack
    stack build
    stack exec ksp-automation

The application currently assumes that kRPC is running on ports 50000 & 50001
of the local machine.

Use `j`/`k` to scroll the mission list & log, `Enter` to start a mission,
`Ctrl-n` & `Ctrl-p` to cycle through the sections.


## Contribute

There are many `TODO`s littered throughout the source, any of these can be
tackled. Ideally they would be moved into Github issues.

For changes with a larger scope, please open an issue for discussion first.


## License

GPL-3.0


[krpc]: https://github.com/krpc/krpc/
[stack]: https://github.com/commercialhaskell/stack/


ChangeLog.md
# Changelog for ksp-automation

## Unreleased changes


src/KSP/Automation.hs
{-# LANGUAGE OverloadedStrings #-}
{-| Utilities for automating Kerbal Space Program using the kRPC mod.

TODO Functionality:

* Parameterized Tasks(set LKO Apoapsis)
* Function to warp to & execute an manually placed maneuver node
* Function to match the inclination of a target
* Hoffman Transfer
* Rendezvous
* Docking
* Landing

TODO Tasks:

* KSC -> LKO -> Dock w/ KerbStation
* Minmus Surface -> Dock w/ MunStation -> Dump Fuel -> Land on Minmus
* Minmus -> Kerbin/Mun Fuel Runs

-}
module KSP.Automation where

import Control.Concurrent (threadDelay)
import Control.Monad (forM_, when, void)
import Control.Monad.IO.Class (liftIO)
import Data.IORef (newIORef, modifyIORef, readIORef)

import KRPCHS
import KRPCHS.SpaceCenter

import qualified Data.Text as T


-- | A function used by Automation programs to log messages.
type LoggingFunction = (T.Text -> RPCContext ())

-- | Log Messages to StdOut
logStdOut :: LoggingFunction
logStdOut =
    liftIO . putStrLn . T.unpack


-- | Used to store the last data received from `liftOff`'s RPC streams.
data LiftOffData
    = LiftOffData
        { lodVelocity :: Double -- ^ Vessel's Surface Velocity
        , lodAltitude :: Double -- ^ Flight's Mean Altitude
        , lodApoapsis :: Double -- ^ Orbit's Apoapsis Altitude
        , lodThrust :: Float    -- ^ Vessel's Current Thrust
        }

-- | An automted routine for taking off & attempting to achieve
-- a circular Low Kerbin Orbit(75,000m).
--
-- This is incomplete and has only been tested on a handful of rocket.
--
-- It starts by firing SRBs & pitching down when various surface velocities
-- & apoapses are reached.
--
-- When the SRBs are depleted it will activate the LF stage and maintain
-- a TWR of ~2.5 until the target apoapsis height is achieved. Then
-- a circularizing manuever node is planned and executed.
--
-- TODO: Eventually attempt to match inclination w/ a target?
lowKerbinOrbit :: LoggingFunction -> StreamClient -> RPCContext ()
lowKerbinOrbit logMessage streamClient = do
    logMessage "Low Kerbin Orbit Program Initiated."
    logMessage "Initializing Vessel Data."
    v <- getActiveVessel
    ctrl <- getVesselControl v
    liftOff logMessage streamClient v
    setControlSAS ctrl True >> setControlSASMode ctrl SASMode'Prograde
    circularizeOrbit logMessage streamClient v
    enableStabilityAssist ctrl
    logMessage "Low Kerbin Orbit Achieved."


-- | Lift-off from the KSC Launchpad & attain an apoapsis of 75km.
liftOff :: LoggingFunction -> StreamClient -> Vessel -> RPCContext ()
liftOff logMessage streamClient v = do
    logMessage "Initializing Lift-Off Data & Telemetry Streams."
    pilot <- getVesselAutoPilot v
    ctrl <- getVesselControl v
    body <- getVesselOrbit v >>= getOrbitBody
    surfaceReferenceFrame <- getCelestialBodyReferenceFrame body
    flight <- vesselFlight v surfaceReferenceFrame

    dataRef <- liftIO . newIORef $ LiftOffData 0 0 0 0
    altitudeStream <- getFlightMeanAltitudeStream flight
    velocityStream <- vesselVelocityStream v surfaceReferenceFrame
    thrustStream <- getVesselThrustStream v
    apoapsisStream <- getVesselOrbit v >>= getOrbitApoapsisAltitudeStream

    logMessage "Setting Launch Parameters."
    setAutoPilotSAS pilot True
    setAutoPilotSASMode pilot SASMode'StabilityAssist
    autoPilotTargetPitchAndHeading pilot 90 90
    setAutoPilotTargetRoll pilot 90
    autoPilotEngage pilot
    setControlThrottle ctrl 1

    countdown logMessage 3
    logMessage "Launch!"
    void $ controlActivateNextStage ctrl

    logMessage "Raising Apoapsis to 75km."
    loop $ do
        -- TODO: Use StateT instead of IORef
        let update = liftIO . modifyIORef dataRef
        -- Update the Data Using Message Streams
        msg <- getStreamMessage streamClient
        when (messageHasResultFor altitudeStream msg) $ do
            altitude <- getStreamResult altitudeStream msg
            update $ \c -> c { lodAltitude = altitude }
        when (messageHasResultFor velocityStream msg) $ do
            velocity <- mag <$> getStreamResult velocityStream msg
            update $ \c -> c { lodVelocity = velocity }
        when (messageHasResultFor thrustStream msg) $ do
            thrust <- getStreamResult thrustStream msg
            update $ \c -> c { lodThrust = thrust }
        when (messageHasResultFor apoapsisStream msg) $ do
            apoapsis <- getStreamResult apoapsisStream msg
            update $ \c -> c { lodApoapsis = apoapsis }

        -- Control Pitch Angle
        liftIO (readIORef dataRef) >>= updatePitch pilot

        -- Decouple SRBs when Empty
        currentStage <- getControlCurrentStage ctrl
        resources <- vesselResourcesInDecoupleStage v currentStage False
        solidFuel <- resourcesAmount resources "SolidFuel"
        liquidFuel <- resourcesAmount resources "LiquidFuel"
        when (solidFuel == 0 && liquidFuel == 0) . void
            $ logMessage "Resources Empty, Decoupling Stage."
            >> controlActivateNextStage ctrl

        -- Adjust Throttle to Maintain 2.5 TWR
        thrust <- liftIO $ lodThrust <$> readIORef dataRef
        weight <- lodAltitude <$> liftIO (readIORef dataRef) >>= weightAtAltitude v
        let twr = thrust / weight
        -- TODO: Fine-tune TWR range using Altitude/Atmosphere Density
        -- TODO: Smooth out changes, very jumpty ATM(see PID Controllers)
        when (twr < 2.3) $
            getControlThrottle ctrl
            >>= (\t -> setControlThrottle ctrl . max 1 $ t + 0.0005)
        when (twr > 2.7) $
            getControlThrottle ctrl
            >>= (\t -> setControlThrottle ctrl . min 0 $ t - 0.0005)

        -- Stop at Apoapsis of 75km
        (> 75000) . lodApoapsis <$> liftIO (readIORef dataRef)

    logMessage "Desired Apoapsis Altitude Achieved, Disabling Throttle."
    setControlThrottle ctrl 0
    autoPilotDisengage pilot
    where
        -- TODO: Smooth out pitch changes, PID Controller?
        updatePitch pilot data_ =
            let
                angles =
                    [ ( ( 0, 0 ), 90 )
                    , ( ( 100, 6500 ), 80 )
                    , ( ( 250, 10000 ), 65 )
                    , ( ( 450, 15000 ), 55 )
                    , ( ( 500, 30000 ), 45 )
                    , ( ( 600, 45000 ), 20 )
                    , ( ( 800, 65000 ), 0 )
                    ]
                velocity = lodVelocity data_
                apoapsis = lodApoapsis data_
            in
                loopMatch angles (\(v', a) -> v' < velocity && a >= apoapsis) (setAutoPilotTargetPitch pilot)


-- | Circularize the current orbit at it's apoapsis.
circularizeOrbit :: LoggingFunction -> StreamClient -> Vessel -> RPCContext ()
circularizeOrbit logMessage streamClient v = do
    pilot <- getVesselAutoPilot v
    ctrl <- getVesselControl v
    -- Calculate DeltaV Required for Circularization, Add Manuever Node
    logMessage "Calculating DeltaV & Burn Time for Circularization."
    (mu, r, a1) <- getVesselOrbit v >>= \orbit ->
        (,,)
            <$> (getOrbitBody orbit >>= getCelestialBodyGravitationalParameter)
            <*> (realToFrac <$> getOrbitApoapsis orbit)
            <*> (realToFrac <$> getOrbitSemiMajorAxis orbit)
    let a2 = r
        v1 = sqrt $ mu * ((2 / r) - (1 / a1))
        v2 = sqrt $ mu * ((2 / r) - (1 / a2))
        deltaV = (v2 - v1) * 1.05
    nodeUT <- getUT
    timeToApoapsis_ <- getVesselOrbit v >>= getOrbitTimeToApoapsis
    node <- controlAddNode ctrl (nodeUT + timeToApoapsis_) deltaV 0 0
    logMessage "Added Circularization Manuever Node."


    -- Calculate Required Burn Time
    -- TODO: Eventually, should account for having to decouple during a burn
    maxThrust <- getVesselAvailableThrust v
    isp <- (* 9.82) <$> getVesselSpecificImpulse v -- TODO: Use current val based on height
    m0 <- getVesselMass v
    let m1 = m0 / exp (deltaV / isp)
        flowRate = maxThrust / isp
        burnTime = (m0 - m1) / flowRate
    logMessage . T.pack $ "Calculated Burn Time of " ++ show (round burnTime :: Integer) ++ "s"


    logMessage "Warping to Approximate Burn Time."
    burnUT <-
        (\t tta -> t + tta - realToFrac burnTime / 2)
        <$> getUT  <*> (getVesselOrbit v >>= getOrbitTimeToApoapsis)
    warpTo (burnUT - 2) 4 10
    logMessage "Warp Completed, Orienting Ship to Burn Direction."


    -- Align to Burn Direction
    nodeFrame <- getNodeReferenceFrame node
    autoPilotEngage pilot
    setAutoPilotReferenceFrame pilot nodeFrame
    setAutoPilotTargetDirection pilot (0, 1, 0)
    logMessage "Ship Oriented for Circularization Burn."


    logMessage "Waiting for Burn Point."
    timeToApoapsisStream <- getVesselOrbit v >>= getOrbitTimeToApoapsisStream
    timeToApoapsisRef <- liftIO $ newIORef 9001
    loop $ do
        msg <- getStreamMessage streamClient
        when (messageHasResultFor timeToApoapsisStream msg) $ do
            timeToApoapsis <- getStreamResult timeToApoapsisStream msg
            void . liftIO . modifyIORef timeToApoapsisRef $ const timeToApoapsis
        (\tta -> tta - realToFrac burnTime / 2 > 0) <$> liftIO (readIORef timeToApoapsisRef)


    -- Start Burn
    logMessage "Beginning Circularization Burn."
    setControlThrottle ctrl 1

    burnVectorStream <- nodeRemainingBurnVectorStream node nodeFrame
    burnVectorRef <- liftIO $ newIORef (0, 9001, 0)
    loop $ do
        msg <- getStreamMessage streamClient
        when (messageHasResultFor burnVectorStream msg) $
            getStreamResult burnVectorStream msg
            >>= void . liftIO . modifyIORef burnVectorRef . const
        burnVector <- liftIO (readIORef burnVectorRef)
        setAutoPilotTargetDirection pilot burnVector
        let (_, progradeVector, _) = burnVector
        when (progradeVector >= 20) $ setControlThrottle ctrl 1
        when (progradeVector < 20 && progradeVector >= 5) $ setControlThrottle ctrl 0.5
        when (progradeVector < 5) $ setControlThrottle ctrl 0.25
        return $ progradeVector < 0
    setControlThrottle ctrl 0

    logMessage "Burn Completed. Removing Manuever Node."
    nodeRemove node
    autoPilotDisengage pilot


-- UTILTIES

-- | Calculate a `Vessel`'s weight at a specific altitude of the
-- `CelestialBody` the vessel is orbiting.
weightAtAltitude :: Vessel -> Double -> RPCContext Float
weightAtAltitude v altitude = do
        vesselMass <- getVesselMass v
        (surfaceGravity, bodyRadius) <-
           getVesselOrbit v >>= getOrbitBody >>= \body ->
            (,) <$> getCelestialBodySurfaceGravity body
                <*> getCelestialBodyEquatorialRadius body
        let surfaceWeight =
                surfaceGravity * vesselMass
            gravityRatio =
                (bodyRadius / (bodyRadius + realToFrac altitude)) ^^ (2 :: Integer)
        return $ surfaceWeight * gravityRatio


-- | Enable SAS & set it to Stability Assist Mode. Cannot be used while
-- autopilot is enabled.
enableStabilityAssist :: Control -> RPCContext ()
enableStabilityAssist ctrl =
    setControlSAS ctrl True >> setControlSASMode ctrl SASMode'StabilityAssist


-- | Determine the Magnitude of a Vector.
mag :: (Double, Double, Double) -> Double
mag (x, y, z) =
    sqrt $ x ^^ (2 :: Integer) + y ^^ (2 :: Integer) + z ^^ (2 :: Integer)


-- | Log a countdown.
countdown :: LoggingFunction -> Int -> RPCContext ()
countdown logMessage n =
    forM_ (reverse [1 .. n]) $ \sec ->
        logMessage (T.pack $ show sec) >> liftIO (threadDelay oneSecond)


-- | Loop an action until it returns True.
loop :: Monad m => m Bool -> m ()
loop a =
    a >>= \b -> if b then return () else loop a


-- | Loop through a list until a match is found, then use it to run an
-- action.
loopMatch :: Monad m => [(a, b)] -> (a -> Bool) -> (b -> m ()) -> m ()
loopMatch matches p f =
    case matches of
        [] ->
            return ()
        (k, v) : rest ->
            if p k then
                f v
            else
                loopMatch rest p f


-- | The `Int` representation of a second for the `threadDelay` function.
--
-- This lets you do things like:
--
-- >>> threadDelay (3 * oneSecond) >>= print    -- Wait 3 Seconds
-- ()
oneSecond :: Int
oneSecond =
    1000000


app/Main.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}
module Main where

import Brick
import Brick.BChan (newBChan, writeBChan)
import Control.Exception (SomeException, catch, bracket)
import Control.Monad (forever)
import Control.Monad.IO.Class (liftIO)
import Control.Concurrent (threadDelay, forkIO, killThread)
import Control.Concurrent.Async (Async, async, cancel)
import Control.Concurrent.MVar (MVar, newMVar, modifyMVar, modifyMVar_)
import Data.Maybe (listToMaybe)
import Data.Time.Clock (getCurrentTime)
import Data.Time.Format (formatTime, defaultTimeLocale)
import KRPCHS (withRPCClient, withStreamClient, runRPCProg)

import KSP.Automation

import qualified Brick.Widgets.Border as B
import qualified Brick.Widgets.Border.Style as B
import qualified Brick.Widgets.Center as C
import qualified Brick.Widgets.List as L
import qualified Data.Text as T
import qualified Data.Vector as Vec
import qualified Graphics.Vty as V


main :: IO ()
main =
        bracket startTick (killThread . snd) (runUI . fst)
    where
        -- Create a `BChan` in a new thread that produces a `Tick` every 250ms
        startTick = do
            chan <- newBChan 10
            tickThreadId <- forkIO . forever $ do
                writeBChan chan Tick
                threadDelay 250000
            return (chan, tickThreadId)
        -- Run the UI loop, cleanup the async Mission thread on exit
        -- TODO: Use MVar to cleanup async thread on main thread exception?
        runUI tickChannel =
            initialState
                >>= customMain (V.mkVty V.defaultConfig) (Just tickChannel) appConfig
                >>= maybe (return ()) cancel . asMissionThread


-- DATA

-- Widget Names
data Name
    = MissionList
    | LogList
    deriving (Eq, Ord, Show)

-- Application Events
data Tick
    = Tick


-- The app's Brick configuration.
appConfig :: App AppState Tick Name
appConfig =
    App
        { appDraw = renderApp
        , appChooseCursor = const listToMaybe
        , appHandleEvent = handleEvents
        , appStartEvent = return
        , appAttrMap = const styleMap
        }

-- The state used within the Brick app.
data AppState =
    AppState
        { asFocusedSection :: AppSection
        , asMissionLog :: L.List Name LogEntry
        , asMissionInProgress :: Bool
        , asEnableAutoStaging :: Bool
        , asMissionList :: L.List Name Mission
        , asMissionThread :: Maybe (Async ())
        , asLogEntriesMVar :: MVar [LogEntry]
        }

initialState :: IO AppState
initialState = do
    loggerMVar <- newMVar []
    logEntry <- makeLogEntry "KSP Automation Initialized - Awaiting Mission Selection."
    return AppState
        { asFocusedSection = MissionSelect
        , asMissionLog = L.list LogList (Vec.fromList [logEntry]) 1
        , asMissionInProgress = False
        , asEnableAutoStaging = False
        , asMissionList = L.list MissionList (Vec.fromList [minBound .. maxBound]) 1
        , asMissionThread = Nothing
        , asLogEntriesMVar = loggerMVar
        }

-- The different sections of the UI the user can cycle through
data AppSection
    = MissionSelect
    | Parameters
    | FlightLog
    | Options
    deriving (Eq, Enum, Bounded)

instance Show AppSection where
    show = \case
        MissionSelect ->
            "Mission Selection"
        Parameters ->
            "Parameters"
        FlightLog ->
            "Flight Log"
        Options ->
            "Options"

-- | The mission programs available.
data Mission
    = LowKerbinOrbit
    | LiftOff
    | Circularize
    | ExecuteManeuver
    deriving (Enum, Bounded)

instance Show Mission where
    show = \case
        LowKerbinOrbit ->
            "Launch to Low Kerbin Orbit"
        LiftOff ->
            "Launchpad Lift-Off"
        Circularize ->
            "Circularize Orbit at Apoapsis"
        ExecuteManeuver ->
            "Execute Planned Maneuver"


data LogEntry
    = LogEntry
        { leDate :: T.Text
        , leMessage :: T.Text
        }

makeLogEntry :: T.Text -> IO LogEntry
makeLogEntry msg = do
    currentTime <- getCurrentTime
    return LogEntry
        { leDate = T.pack $ formatTime defaultTimeLocale "%T" currentTime
        , leMessage = msg
        }


-- STYLES

defaultStyle :: AttrName
defaultStyle =
    attrName "default"

logBrackets :: AttrName
logBrackets =
    attrName "logBrackets"

logDate :: AttrName
logDate =
    attrName "logDate"

logMessage :: AttrName
logMessage =
    attrName "logMessage"

inputLabelAttr :: AttrName
inputLabelAttr =
    attrName "inputLabel"

focusedBorder :: AttrName
focusedBorder =
    attrName "focusedBorder"

focusedActiveListItem :: AttrName
focusedActiveListItem =
    attrName "focusedActiveListItem"

unfocusedActiveListItem :: AttrName
unfocusedActiveListItem =
    attrName "unfocusedActiveListItem"

focusedListItem :: AttrName
focusedListItem =
    attrName "focusedListItem"

unfocusedListItem :: AttrName
unfocusedListItem =
    attrName "unfocusedListItem"

focusedBorderColor :: AttrName
focusedBorderColor =
    attrName "focusedBorderColor"

unfocusedBorderColor :: AttrName
unfocusedBorderColor =
    attrName "unfocusedBorderColor"

styleMap :: AttrMap
styleMap =
    attrMap (fg V.magenta)
        [ ( defaultStyle, fg V.magenta )
        -- Form Inputs
        , ( inputLabelAttr, V.withStyle (fg V.magenta) V.bold )
        -- Section Borders
        , ( focusedBorderColor, fg V.magenta )
        , ( unfocusedBorderColor, fg V.white )
        -- Mission List
        , ( focusedActiveListItem, V.withStyle (V.black `on` V.magenta) V.bold )
        , ( unfocusedActiveListItem, V.withStyle (V.white `on` V.magenta) V.bold )
        , ( focusedListItem, fg V.magenta )
        , ( unfocusedListItem, fg V.white )
        -- Flight Log
        , ( logBrackets, fg V.white )
        , ( logDate, fg V.magenta )
        , ( logMessage, fg V.white )
        ]


-- RENDER

renderApp :: AppState -> [Widget Name]
renderApp s =
    let
        mainWindow =
                renderLeft s
            <+> renderRight s
            <=> str "STATUS BAR"
    in
        [ mainWindow
        ]


renderLeft :: AppState -> Widget Name
renderLeft s =
    let
        longestMissionText =
            maximum
                $ map (textWidth . show) [minBound .. maxBound :: Mission]
    in
        vBox
            [ renderSection MissionSelect s
                . hLimit (longestMissionText + 4)
                $ renderMissionList s
            , renderSection Options s
                $ renderOptions s
            ]


renderMissionList :: AppState -> Widget Name
renderMissionList s =
    let
        isFocusedSection =
            asFocusedSection s == MissionSelect
    in
        L.renderList renderMission isFocusedSection (asMissionList s)
    where
        renderMission isSelected mission =
            let
                attribute
                    | isSelected && asFocusedSection s == MissionSelect =
                        withAttr focusedActiveListItem
                    | isSelected =
                        withAttr unfocusedActiveListItem
                    | asFocusedSection s == MissionSelect  =
                        withAttr focusedListItem
                    | otherwise =
                        withAttr unfocusedListItem
            in
                 attribute
                    . padLeft (Pad 1)
                    . padRight Max
                    . str
                    $ show mission


-- TODO; Implement Form Fields
renderOptions :: AppState -> Widget Name
renderOptions _ =
    padTopBottom 1 . padLeftRight 2 $ vBox
        [ hBox
            [ withAttr inputLabelAttr
                . padRight (Pad 3)
                $ str "Enable Auto-Staging:"
            , str "[ X ]"
            ]
        ]


renderRight :: AppState -> Widget Name
renderRight s =
        renderSection Parameters s (renderMissionParameters s )
    <=> renderSection FlightLog s (renderMissionLog (asMissionLog s))


-- TODO: Implement Brick.Form Fields
renderMissionParameters :: AppState -> Widget Name
renderMissionParameters _ =
    padTopBottom 1 $ C.hCenter $ hBox
        [ withAttr inputLabelAttr $ padRight (Pad 2) $ vBox
            [ str "Target Apoapsis(meters):"
            , str "Match Target Inclination:"
            ]
        , vBox
            [ str "[ 75,000 ]"
            , str "     [ X ]"
            ]
        ]


renderMissionLog :: L.List Name LogEntry -> Widget Name
renderMissionLog =
        L.renderList renderEntry False
    where
        renderEntry _ entry =
            padRight Max $
            withAttr logBrackets (str "[")
                <+> withAttr logDate (txt $ leDate entry)
                <+> withAttr logBrackets (str "] ")
                <+> withAttr logMessage (txtWrap $ leMessage entry)


renderSection :: AppSection -> AppState -> Widget n -> Widget n
renderSection section state =
    let
        isFocused =
            asFocusedSection state == section
        borderLabel =
            B.borderWithLabel (str $ " " ++ show section ++ " ")
                . withDefAttr defaultStyle
    in
        if isFocused then
            withDefAttr focusedBorderColor
                . withBorderStyle B.unicodeBold
                . borderLabel
        else
            withDefAttr unfocusedBorderColor
                . borderLabel


-- UPDATE

-- TODO: This should just dispatch to other functions
handleEvents :: AppState -> BrickEvent Name Tick -> EventM Name (Next AppState)
handleEvents s = \case
    -- Add Log Entries from Shared MVar to AppState on Every Tick
    AppEvent Tick -> do
        -- TODO: Poll async mission thread & update state
        newLogEntries <- liftIO . modifyMVar (asLogEntriesMVar s)
            $ \entries -> return ([], reverse entries)
        continue s
            { asMissionLog =
                foldl (flip $ L.listInsert maxBound) (asMissionLog s) newLogEntries
            }

    VtyEvent ev -> do
        let focusedSection =
                asFocusedSection s
        -- TODO: Handle Log Scrolling
        updatedState <-
            case focusedSection of
                MissionSelect ->
                    (\l -> s { asMissionList = l })
                        <$> L.handleListEventVi L.handleListEvent ev (asMissionList s)
                FlightLog ->
                    (\l -> s { asMissionLog = l })
                        <$> L.handleListEventVi L.handleListEvent ev (asMissionLog s)
                _ ->
                    return s
        -- TODO: Split each section's keybindings into separate function
        case ev of
            -- Launch Mission
            V.EvKey V.KEnter [] ->
                let
                    maybeSelectedMission =
                        fmap snd . L.listSelectedElement $ asMissionList updatedState
                    runAsyncMission =
                        liftIO . async . maybe (return ()) (runMission $ asLogEntriesMVar s)
                in
                    if focusedSection == MissionSelect then
                        runAsyncMission maybeSelectedMission
                            >>= (\asyncProc -> continue updatedState { asMissionThread = Just asyncProc })
                    else
                        continue updatedState

            -- Quit
            V.EvKey (V.KChar 'q') [] ->
                halt updatedState
            V.EvKey V.KEsc [] ->
                halt updatedState

            -- Cycle through Sections
            V.EvKey (V.KChar 'n') [V.MCtrl] ->
                continue $ updatedState
                    { asFocusedSection = nextEnum focusedSection }
            V.EvKey (V.KChar 'p') [V.MCtrl] ->
                continue $ updatedState
                    { asFocusedSection = previousEnum focusedSection }
            _ ->
                continue updatedState
    _ ->
        continue s


runMission  :: MVar [LogEntry] -> Mission -> IO ()
runMission logMVar m =
    let
        logger =
            liftIO . messageLogger

        prog =
            case m of
                LowKerbinOrbit ->
                    lowKerbinOrbit
                _ ->
                    fail "UNIMPLEMENTED MISSION"
        run =
            withRPCClient (show m) "127.0.0.1" "50000" $ \client ->
            withStreamClient client "127.0.0.1" "50001" $ \streamClient ->
                runRPCProg client (prog logger streamClient)
    in
        catch run $ \(e :: SomeException) ->
            messageLogger . T.pack $ "Mission Program Exception!\n" ++ show e
    where
        messageLogger :: T.Text -> IO ()
        messageLogger msg = do
            newEntry <- makeLogEntry msg
            modifyMVar_ logMVar (return . (:) newEntry)


nextEnum :: (Enum a, Bounded a, Eq a) => a -> a
nextEnum e =
    if e == maxBound then
        minBound
    else
        succ e

previousEnum :: (Enum p, Bounded p, Eq p) => p -> p
previousEnum e =
    if e == minBound then
        maxBound
    else
        pred e
